Migrate more RemoteLayerTreeTransaction members from transaction scope to main frame scope

ruthvikkonda opened the pull request:
#### 5f8ac173285d5ec1550606ddc8383c45966e7f94
<pre>
Migrate more RemoteLayerTreeTransaction members from transaction scope to main frame scope
<a href="https://bugs.webkit.org/show_bug.cgi?id=301305">https://bugs.webkit.org/show_bug.cgi?id=301305</a>
<a href="https://rdar.apple.com/163225069">rdar://163225069</a>

Reviewed by Matt Woodrow.

Right now RemoteLayerTreeDrawingAreaProxy::commitLayerTree currently receives a Vector&lt;std::pair&lt;RemoteLayerTreeTransaction,
RemoteScrollingCoordinatorTransaction&gt;&gt;, one pair per root frame. However, some aspects of the transaction are
scoped to the root frame. Some are scoped to the page. And some are scoped to the main frame. We want to accordingly
introduce some separation/organization here to reduce duplication and unnecessary isMainFrameProcessTransaction() checks.

This PR builds on <a href="https://github.com/WebKit/WebKit/pull/52671.">https://github.com/WebKit/WebKit/pull/52671.</a> In this PR, we move more relevant members from transaction
scope to main frame scope and page scope. They get accordingly removed from the isMainFrameProcessTransaction() checks
on UIProcess side and handled separately. Additionally, we introduce a new PageData struct and add a member to it.

Once we get toward the end of the migration and almost all mainframe/page members are moved out of transaction level
code paths, we can introduce more new methods that are only called once for processing PageData/MainFrameData
rather than using existing methods that get called for every transaction. But for now, we leave most of them as part of
the same methods to reduce breakage.

We also add a MESSAGE_CHECK at the start of commitLayerTree() to verify that main frame data is only provided
from the main frame process.

No new functionality is added. This is a simple refactoring. Existing test coverage is enough.
As a result, no new tests are added.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeCommitBundle.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
(WebKit::RemoteLayerTreeTransaction::themeColor const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setThemeColor): Deleted.
(WebKit::RemoteLayerTreeTransaction::pageExtendedBackgroundColor const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setPageExtendedBackgroundColor): Deleted.
(WebKit::RemoteLayerTreeTransaction::sampledPageTopColor const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setSampledPageTopColor): Deleted.
(WebKit::RemoteLayerTreeTransaction::pageScalingLayerID const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setPageScalingLayerID): Deleted.
(WebKit::RemoteLayerTreeTransaction::scrolledContentsLayerID const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setScrolledContentsLayerID): Deleted.
(WebKit::RemoteLayerTreeTransaction::mainFrameClipLayerID const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setMainFrameClipLayerID): Deleted.
(WebKit::RemoteLayerTreeTransaction::isInStableState const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setIsInStableState): Deleted.
(WebKit::RemoteLayerTreeTransaction::callbackIDs const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setCallbackIDs): Deleted.
(WebKit::RemoteLayerTreeTransaction::newlyReachedPaintingMilestones const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setNewlyReachedPaintingMilestones): Deleted.
(WebKit::RemoteLayerTreeTransaction::hasEditorState const): Deleted.
(WebKit::RemoteLayerTreeTransaction::editorState const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setEditorState): Deleted.
(WebKit::RemoteLayerTreeTransaction::fixedContainerEdges const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setFixedContainerEdges): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::description const):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _didCommitLayerTree:mainFrameData:]):
(-[WKWebView _didCommitLayerTree:]): Deleted.
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::didCommitLayerTree):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::didCommitLayerTree):
(WebKit::WebPageProxy::didCommitMainFrameData):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
(WebKit::RemoteLayerTreeDrawingAreaProxy::didCommitLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::updateBannerLayers):
(WebKit::RemoteLayerTreeHost::updateLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::didCommitLayerTree):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLayerTree):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didCommitLayerTree):
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _didCommitLayerTree:mainFrameData:]):
(-[WKContentView _didCommitLayerTree:]): Deleted.
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::willCommitLayerTree):
(WebKit::WebPage::willCommitMainFrameData):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
(WebKit::RemoteLayerTreeDrawingArea::willCommitLayerTree):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.h:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::willCommitLayerTree):

Canonical link: <a href="https://commits.webkit.org/302408@main">https://commits.webkit.org/302408@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/08ebc1929159242125f1cd44bc010b3e2e8e7686

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128963 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1216 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39794 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136342 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80321 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9978cb13-395a-45f0-9909-e61c800a96af) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1156 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1095 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98183 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66098 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7c90cb89-477f-4e99-9549-22395fbf2097) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131910 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/889 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115524 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78811 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/61f2a827-faea-449a-8efd-471fa1df97ce) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/816 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79623 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109258 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34134 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138809 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1017 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/992 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106761 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1078 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111862 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106538 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27132 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/846 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30383 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53491 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1093 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64486 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/927 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/979 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1024 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
```diff
@@ -282,12 +282,6 @@ static void dumpChangedLayers(TextStream& ts, const LayerPropertiesMap& changedL
     if (m_pageScaleFactor != 1)
         ts.dumpProperty("pageScaleFactor"_s, m_pageScaleFactor);
 
-#if PLATFORM(MAC)
```
Not super urgent, but it'd be nice to be able to log the Page/MainFrame objects (there's `RemoteLayerTree` logging that prints the transactions as they arrive currently).

ruthvikkonda replied:
Sounds good. I'll take care of this in the next PR.

mattwoodrow commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -317,12 +317,40 @@ - (void)drawInContext:(CGContextRef)cgContext
             m_activityStateChangeForUnhidingContent = std::nullopt;
         }
+
+        RefPtr page = this->page();
```
Can we add a MESSAGE_CHECK that `connection` is the WebPageProxy's mainFrameProcess's connection?

We should only accept `mainFrameData` from that process.

mattwoodrow replied:
Actually, adding the same check on `isMainFrameTransaction` would be nice too.

ruthvikkonda replied:
Sure. Rather than add it at every call site of `bundle.mainFrameData` and every call site of `isMainFrameTransaction()`, I think it'd be better if we just validated once at function entry. This is a simpler, more maintainable approach that allows for malicious processes to get killed immediately before any processing and creates a single validation point. So it'd be something like

```
void RemoteLayerTreeDrawingAreaProxy::commitLayerTree(IPC::Connection& connection, const RemoteLayerTreeCommitBundle& bundle, HashMap<ImageBufferSetIdentifier, std::unique_ptr<BufferSetBackendHandle>>&& handlesMap)
{
    bool hasMainFrameProcessTransaction { false };
    for (auto& transaction : bundle.transactions) {
        if (transaction.isMainFrameProcessTransaction())
            hasMainFrameProcessTransaction = true;
            break;
    }

    if (bundle.mainFrameData || hasMainFrameProcessTransaction)
        MESSAGE_CHECK_BASE()
    
    .....rest of code
```

Does this sound good to you?

mattwoodrow replied:
Yes please :)

ruthvikkonda replied:
Done! Hopefully that change looks right to you.

ruthvikkonda committed:
Migrate more RemoteLayerTreeTransaction members from transaction scope to main frame scope
https://bugs.webkit.org/show_bug.cgi?id=301305
rdar://163225069

Reviewed by Matt Woodrow.

Right now RemoteLayerTreeDrawingAreaProxy::commitLayerTree currently receives a Vector<std::pair<RemoteLayerTreeTransaction,
RemoteScrollingCoordinatorTransaction>>, one pair per root frame. However, some aspects of the transaction are
scoped to the root frame. Some are scoped to the page. And some are scoped to the main frame. We want to accordingly
introduce some separation/organization here to reduce duplication and unnecessary isMainFrameProcessTransaction() checks.

This PR builds on https://github.com/WebKit/WebKit/pull/52671. In this PR, we move more relevant members from transaction
scope to main frame scope and page scope. They get accordingly removed from the isMainFrameProcessTransaction() checks
on UIProcess side and handled separately. Additionally, we introduce a new PageData struct and add a member to it.

Once we get toward the end of the migration and almost all mainframe/page members are moved out of transaction level
code paths, we can introduce more new methods that are only called once for processing PageData/MainFrameData
rather than using existing methods that get called for every transaction. But for now, we leave most of them as part of
the same methods to reduce breakage.

We also add a MESSAGE_CHECK at the start of commitLayerTree() to verify that main frame data is only provided
from the main frame process.

No new functionality is added. This is a simple refactoring. Existing test coverage is enough.
As a result, no new tests are added.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeCommitBundle.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
(WebKit::RemoteLayerTreeTransaction::themeColor const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setThemeColor): Deleted.
(WebKit::RemoteLayerTreeTransaction::pageExtendedBackgroundColor const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setPageExtendedBackgroundColor): Deleted.
(WebKit::RemoteLayerTreeTransaction::sampledPageTopColor const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setSampledPageTopColor): Deleted.
(WebKit::RemoteLayerTreeTransaction::pageScalingLayerID const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setPageScalingLayerID): Deleted.
(WebKit::RemoteLayerTreeTransaction::scrolledContentsLayerID const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setScrolledContentsLayerID): Deleted.
(WebKit::RemoteLayerTreeTransaction::mainFrameClipLayerID const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setMainFrameClipLayerID): Deleted.
(WebKit::RemoteLayerTreeTransaction::isInStableState const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setIsInStableState): Deleted.
(WebKit::RemoteLayerTreeTransaction::callbackIDs const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setCallbackIDs): Deleted.
(WebKit::RemoteLayerTreeTransaction::newlyReachedPaintingMilestones const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setNewlyReachedPaintingMilestones): Deleted.
(WebKit::RemoteLayerTreeTransaction::hasEditorState const): Deleted.
(WebKit::RemoteLayerTreeTransaction::editorState const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setEditorState): Deleted.
(WebKit::RemoteLayerTreeTransaction::fixedContainerEdges const): Deleted.
(WebKit::RemoteLayerTreeTransaction::setFixedContainerEdges): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::description const):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _didCommitLayerTree:mainFrameData:]):
(-[WKWebView _didCommitLayerTree:]): Deleted.
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::didCommitLayerTree):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::didCommitLayerTree):
(WebKit::WebPageProxy::didCommitMainFrameData):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
(WebKit::RemoteLayerTreeDrawingAreaProxy::didCommitLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::updateBannerLayers):
(WebKit::RemoteLayerTreeHost::updateLayerTree):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::didCommitLayerTree):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLayerTree):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::didCommitLayerTree):
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView _didCommitLayerTree:mainFrameData:]):
(-[WKContentView _didCommitLayerTree:]): Deleted.
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::willCommitLayerTree):
(WebKit::WebPage::willCommitMainFrameData):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
(WebKit::RemoteLayerTreeDrawingArea::willCommitLayerTree):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.h:
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::willCommitLayerTree):

Canonical link: https://commits.webkit.org/302408@main
